Rewrite

forensicmatt opened the pull request:
Refactor of code.

forensicmatt committed:
refactor

complete refactor;

forensicmatt committed:
added search size

separated chunk size and search size;

forensicmatt committed:
debug offset

debug offset

forensicmatt committed:
regex fix

regex fix

forensicmatt committed:
added multithreading

parallel parsing. Thanks to https://github.com/omerbenamram/evtx for good examples.

forensicmatt committed:
increased chunk size

increased chunk and search sizes for better performance;

forensicmatt committed:
record validation

added record validation which helps reduce false positives mainly for caving extracted unallocated data;

forensicmatt committed:
record validation

added record validation which helps reduce false positives mainly for caving extracted unallocated data;

forensicmatt committed:
Regex Optimization

Use lazy_static for Regex

omerbenamram commented on src/timestamp.rs:
```diff
@@ -0,0 +1,48 @@
+        write!(f,"{}",self.to_datetime())
+    }
+}
+impl ser::Serialize for WinTimestamp {
```
Consider storing `chrono` time objects in your structs, they implement `Serialize`, and since JSON specifies the time format exactly, you get the good behavior for free. 

Also, you could go back to using the implementation from `winstructs` :)

omerbenamram commented on src/usn.rs:
```diff
@@ -0,0 +1,331 @@
+        inner_handle.seek(SeekFrom::Start(0))?;
+
+        Ok( Self {
+            _inner_handle: inner_handle,
```
Contrary to python, the rust convention is to use visibility modifier (e.g `pub` when having public attributes, otherwise private), no need for the _ prefix.

omerbenamram commented on src/usn.rs:
```diff
@@ -0,0 +1,331 @@
+        lazy_static! {
+            static ref RE_USN: bytes::Regex = bytes::Regex::new(
+                "(?-u)..\x00\x00\x02\x00\x00\x00"
+            ).unwrap();
```
Maybe https://github.com/shepmaster/jetscii could make this go even faster :)

omerbenamram replied:
And if you keep regex, it's would have been nicer to use `.expect('This is a valid regex')` to let the reader know that this cannot possibly fail.

forensicmatt replied:
Going to leave Regex in for now because I would like at some point to account for version 3 and 4 which would change regex to `"(?-u)..\x00\x00(\x02|\x03|\x04)\x00\x00\x00"` (or similar). I did remove unwrap in favor of expect though.

omerbenamram commented on src/bin/rusty_usn.rs:
```diff
@@ -0,0 +1,182 @@
+
+fn set_debug_level(matches: &ArgMatches){
+    // Get the possible logging level supplied by the user
+    let message_level = match matches.is_present("debug") {
```
Personally I think `-v, -vv, -vvv` is nicer (and more common), but it's a matter of taste :)

forensicmatt replied:
I will leave this. I like being able to have the one to one with the exact verbage of the logging lib.

omerbenamram commented on src/bin/rusty_usn.rs:
```diff
@@ -0,0 +1,182 @@
+use clap::{App, Arg, ArgMatches};
+use RustyUsn::usn::{UsnParserSettings, UsnParser};
+
+static VERSION: &'static str = "1.0.0";
```
use `env!("CARGO_PKG_VERSION")`

forensicmatt replied:
cool! I didn't know this existed.

omerbenamram commented on src/bin/rusty_usn.rs:
```diff
@@ -0,0 +1,182 @@
+                Some("Debug") => LevelFilter::Debug,
+                Some("Trace") => LevelFilter::Trace,
+                Some(unknown) => {
+                    panic!("Unknown debug level [{}]", unknown);
```
eprintln! and exit could be a little nicer, traceback is not very helpful here.

omerbenamram commented on src/record.rs:
```diff
@@ -0,0 +1,193 @@
+            );
+        }
+
+        let file_reference = MftReference::from_reader(&mut buffer).unwrap();
```
Don't unwrap here, wrap the error from `winstructs` with your own error instead :) (implement From<>..)

omerbenamram reviewed (commented):
Really nice work!

Consider enabling`rustfmt` on save to make code formatting consistent.

forensicmatt committed:
remove unwraps

UsnError from WinstructError to remove unwrap;

forensicmatt committed:
Use chrono DateTime

Use chrono DateTime instead of custom WinTimestamp; Removed custom WinTimestamp;

forensicmatt committed:
updated tests and examples

updated tests and examples

forensicmatt committed:
var name changes

var name changes

forensicmatt committed:
replace unwrap()

replace unwrap() with expect()

forensicmatt committed:
Version from Cargo

Version from Cargo

forensicmatt committed:
Removed Panics

Removed Panics

forensicmatt committed:
Updated README

Updated README

forensicmatt committed:
minor changes;

Updated Cargo info; changed package name;

forensicmatt committed:
updated package name

updated package name; fixed tests;

forensicmatt committed:
azure pipeline

azure pipeline
